### PR TITLE
Adding a menu of functions for testlib

### PR DESCRIPTION
**Pull request by castulo**
This commit adds a menu of the functions that are available for
users when writing tests or just using the library to create
resources to test their code.

Signed-off-by: Castulo Martinez <castulo.martinez@intel.com>

### CODE REVIEW

**Comment by icarus-sparry** [comment]
I don't think the implementation is a good idea. POSIX only allows the 'foo()' format for declaring a function. 'function foo' is an extension which means different things in ksh, bash, dash etc. In dash it is invalid, in bash it is the same and in ksh it introduces different scoping rules for local variables.

I think a better implementation would use a "magic comment" to mark things that should be ignored.

 I wonder if the sorts are a good idea. The fact that you are splitting function names which contain "assert" apart suggests that a pure sorted list is not exactly what is needed.

Perhaps you could instead look for lines matching say '^## [0-9]+ ', sort them, and then remove the first two 'words' would allow a more general arrangement of the output. It would also allow you to remove the special case code for the headings, you could have '## 0 heading for section 1' and '## 2 heading for section 3' lines.

**Comment by otaviobp** [comment]
> I don't think the implementation is a good idea. POSIX only allows the 'foo()' format for declaring a function. 'function foo' is an extension which means different things in ksh, bash, dash etc. In dash it is invalid, in bash it is the same and in ksh it introduces different scoping rules for local variables.

We are using bats for testing and only bash is supported, so I don't see a reason on being concerned about that.

> I think a better implementation would use a "magic comment" to mark things that should be ignored.

Anyway, I think it's a better idea to use a magic comment. But instead of a comment for ignore I would add a comment for listing it. You could have

     foo() { #swupd_function

and 

    assert_bla() { #swupd_assert

and grep for that comment.

> I wonder if the sorts are a good idea. The fact that you are splitting function names which contain "assert" apart suggests that a pure sorted list is not exactly what is needed.

I'm ok on just a regular sort for now as we don't have a really big number of functions. But I agree that it's nice to have that organized by sections

**Comment by castulo** [comment]
Thanks @icarus-sparry and @otaviobp that is really good feedback

File: test/functional/testlib.bash [hunk 1]
```diff
@@ -2693,3 +2705,40 @@ assert_files_not_equal() {
...
+# Parameters:
+# - prefix: a prefix that could be used to filter functions
+testlib() {
+
+	local prefix=$1
+	local funcs
+	local assertions
+
+	funcs=$(grep "$prefix.*() { [#] swupd_function" "$FUNC_DIR"/testlib.bash | cut -f1 -d "(")
+	assertions=$(grep "$prefix.*() { [#] assertion" "$FUNC_DIR"/testlib.bash | cut -f1 -d "(")
```
Comment (jwakre) [inline test/functional/testlib.bash]: For these regex to match the function prefixes, they should start at the beginning of the line.

File: test/functional/testlib.bash [hunk 2]
```diff
@@ -2693,3 +2705,40 @@ assert_files_not_equal() {
...
+	echo ""
+	if [ -z "$prefix" ]; then
```
Comment (otaviobp) [inline test/functional/testlib.bash]: I prefer to leave the header even if prefix is set.

File: test/functional/testlib.bash (hunk 1 above)
Comment (otaviobp) [reply test/functional/testlib.bash]: You are right. But I prefer the way it is now.
So:

     $ testlib not

will print all functions with not in the name, like:

    assert_status_is_not
    assert_dir_not_exists
    assert_file_not_exists
    assert_not_in_output
    assert_is_not_output
    assert_regex_not_in_output
    assert_regex_is_not_output
    assert_not_equal
    assert_files_not_equal

Comment (jwakre) [reply test/functional/testlib.bash]: In that case, I think the prefix argument name and description should be changed to remove confusion.

Comment (castulo) [reply test/functional/testlib.bash]: that is a good idea @jwakre, I'll update that

File: test/functional/testlib.bash (hunk 2 above)
Comment (castulo) [reply test/functional/testlib.bash]: I will fix that @otaviobp

**Comment by castulo** [comment]
@otaviobp, @jwakre, I have addressed your issues in this new PR update.

**Pull request merged**